### PR TITLE
Blue Green GAP Content

### PR DESCRIPTION
**Pull request by vimorra**
Content dedicated for the Blue Green AKS GAP.
Added a dedicated folder for the blue green scenario that will include also the terraform implementation

### CODE REVIEW

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+The answer that will make everyone happy from business into infra to apps is blue green deployment at Kubernetes infra level.
+With modern principles and availability of cloud services like:
+- [IaC](https://docs.microsoft.com/en-us/devops/deliver/what-is-infrastructure-as-code)
+- [Immutable Infrastructure](https://www.hashicorp.com/resources/what-is-mutable-vs-immutable-infrastructure)
+- [Cloud Elasticity](https://azure.microsoft.com/en-us/overview/what-is-elastic-computing/)
+- [Continuous Delivery](https://docs.microsoft.com/en-us/devops/deliver/what-is-continuous-delivery)
+
+Blue Green deployment has become a de-facto standard pattern for the release management and operation at infra and application level in kubernetes environments.
+
+In this article is described the design and implementation of the blue green deployment for AKS leveraging Azure Cloud managed services and native kubernetes features. Adoption of this pattern improves the operation and availability during the deployment of changes/upgrades of AKS clusters.
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:13]: Consider changing "In this article is described" to "This articles describes"

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+
+## Potential use cases
+This solution is a generalized architecture pattern, which can be used for many different scenarios and industries. In particular it can be applied in AKS deployments and is also used for [Mission Critical scenario](https://docs.microsoft.com/en-us/azure/architecture/reference-architectures/containers/aks-multi-region/aks-multi-cluster). See the following example solutions that build off of this core architecture:
+
+- [Link to first solution idea or other architecture that builds off this solution](filepath.yml)
+
+
+## Architecture
+Below is the high level architecture that describes the pattern and related services invovled. The [Worklfow section](#workflow) describ in detail the steps for the implementation of the pattern and in particular the sequence of events to have the proper switch between the clusters.
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:38]: Change "describ" to "describes".

Also rephrase "the sequence of events to have the proper switch between the clusters." as it is not clear what you mean.

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+It is importnant to highlight that this pattern is like a state machine. The blue or green clusters are both on at the same time, but only for a limitied period of time. This is done to optmize the costs and operational effort.
+
+5 stages of the Pattern:
+1. T0: Blue Cluster is On
+2. T1: Green Cluster Deployment
+3. T2: Sync K8S State between Blue and Green clusters
+4. T3: Traffic Switch to the green cluster
+5. T4: Blue cluster is destroyed
+
+Once implemented, the workflow will start at the next planned release of the cluster beginning with T1: Green Cluster Deployment. The triggers to transition in the multiples stages can be programtically configured and executed, this is usually the desired end state, that start from a manual/semi-automatic implementation. The triggers are related to functional parameters together with SLO/SLI defined at operation levele to cover apps and infra aspects.
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:58]: Please review the first and last sentences in this paragraph as it is not clear what you mean.

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+This pattern is flexible on the network discoverabiity of the clusters, in fact you can have multiple options:
+- A DNS record dedicated to the blue and green clusters IP
+- A DNS record dedicated to the blue green cluster pointing to the App Gateway IP
+
+
+#### T0: Blue Cluster is On
+The initial stage of the pattern is to have the existing live cluster on, which is called the Blue Cluster. At this stage we are preparing for the deployment of the new version of the cluster.
+
+![Step0](../media/bg-step0.png)
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:67]: Do the green App Gateways need to be deployed at this time? Can't we deploy them in the next step?

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+#### T1: Green Cluster Deployment
+Now the deployment of the new version is started and the the new Green Cluster is deployed in parallel to the existing Blue Cluster. After the Green cluster is deployed, the live traffic is still routed in the Blue Cluster. 
+
+![Step1](../media/bg-step1.png)
+
+The trigger to move into the [T2 stage](#t2-sync-k8s-state-between-blue-and-green-cluster) is the validation of AKS at platform level, using native metrics available in Azure Monitor and CLI commands to check the health of the deployment.
+
+
+#### T2: Sync K8S State between Blue and Green cluster
+At this stage there is an parity between the two clusters, meaning all **applications**, **operators** and **K8S resources** are newly deployed in the green cluster. The ultimate goal is that at the end of the sync the clusters are equivalent.
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:81]: If there is a parity between the two clusters, why do we still need to sync them? Maybe consider saying that the two clusters have identical software and configuration deployed by their local state is different.

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+There are multiple solutions/approaches to replicate/sync K8S state on clusters:
+- Redeployment via CI/CD
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:84]: Not sure what this means? What are we redeploying here?

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+- GitOps with solutions promoted in CNCF
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:85]: Any example of solutions that you can link to?

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+- Customized solution that store the K8S configs and resources, usually databases and K8S manifests generators
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:86]: Any link to an example of such a solution?

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+
+![Step2](../media/bg-step2.png)
+
+Usually syncing the deployment of new applications is not permitted in the live cluster, so there is a period of time that starts with the sync and finishs when the switch to the green is completed. This period can be avoided with advanced mechanisms that manage the K8S state in multiple clusters, but that is not part of this article.
+
+When the sync is completed, a test/validation of the cluster is required. This include a check on the monitoring and logging platforms to validate the health of the cluster. Usually the Green cluster is exposed on the App Gateway or External LB with an internal URL, that is not visibile for external users.
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:92]: Change "This include" to "This includes".

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+This test/validation stage is used as a trigger for the [next state](#t3-traffic-switch-to-the-green-cluster) of the pattern.
+
+
+#### T3: Traffic Switch to the green cluster
+After the sync is complete and Green cluster vaidated at platform and application level, then the Green cluster is ready to be promoted as the primary cluster to start receiving the live traffic. The switch is done at the networking level. Often the workloads are stateless. However, if the workloads are stateful then an additional solution must be implemented to mantain the state and caching during the switch.
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:98]: Regarding the last sentence in this paragraph., didn't you mention this already in the previous step T2?

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+![Step0](../media/bg-step0.png)
+
+The trigger condition for the [next stage](#t1-green-cluster-deployment) is the release of a new version of the cluster.
+
+
+#### T1: Green Cluster Deployment
+Now the deployment of the new version is started and the the new Green Cluster is deployed in parallel to the existing Blue Cluster. After the Green cluster is deployed, the live traffic is still routed in the Blue Cluster. 
+
+![Step1](../media/bg-step1.png)
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:75]: The subnets are missing in the diagram. Also the arrows need to be fixed.

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+
+#### T2: Sync K8S State between Blue and Green cluster
+At this stage there is an parity between the two clusters, meaning all **applications**, **operators** and **K8S resources** are newly deployed in the green cluster. The ultimate goal is that at the end of the sync the clusters are equivalent.
+
+There are multiple solutions/approaches to replicate/sync K8S state on clusters:
+- Redeployment via CI/CD
+- GitOps with solutions promoted in CNCF
+- Customized solution that store the K8S configs and resources, usually databases and K8S manifests generators
+
+![Step2](../media/bg-step2.png)
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:88]: Fix the arrows on the diagram

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+- Green Cluster host - the dedicated host for the green cluster
+
+The cluster host is the one configured at Application Gateway level to manage the ingress traffic, moreover the hostname is also part of the AKS Ingress configuration in order to manage the TLS properly. This host is used only for live transactions and requests.
+
+The Blue and Green cluster hosts are mainly used for:
+- Test and Validate the specific cluster, like mentioned in [Step 2](#t2-sync-k8s-state-between-blue-and-green-cluster)
+- The switch of the live traffic, mainly the hosts are used for the backend pools configuration at Application Gateway level, in this way the switch is transparent for the end user of the workloads.
+For the testing and validation purpose the hosts are also exposed at Application Gateway level with dedicated endpoints and also at AKS Ingress controller level to manage the TLS in the proper way.
+At this stage the validation is based on the infra and app monitoring metrics and official SLO and SLI, when avaialble. If the validation gate is satisfied than is possible to move in the [last state](#t4-blue-cluster-is-destroyed) of the pattern.
+
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:116]: Change "than" to "then" in the last sentence.

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+This test/validation stage is used as a trigger for the [next state](#t3-traffic-switch-to-the-green-cluster) of the pattern.
+
+
+#### T3: Traffic Switch to the green cluster
+After the sync is complete and Green cluster vaidated at platform and application level, then the Green cluster is ready to be promoted as the primary cluster to start receiving the live traffic. The switch is done at the networking level. Often the workloads are stateless. However, if the workloads are stateful then an additional solution must be implemented to mantain the state and caching during the switch.
+
+An important point to mention is that this pattern is based on a full switch, meaning 100% of the traffic is routed to the new cluster when the switch is applied. Alternative switching, like canary releases, are possible but not in the scope the article.
+
+![Step3](../media/bg-step3.png)
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:102]: Consider using green arrows between the green components traffic to be consistent with the very first diagram.

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+
+The Blue and Green cluster hosts are mainly used for:
+- Test and Validate the specific cluster, like mentioned in [Step 2](#t2-sync-k8s-state-between-blue-and-green-cluster)
+- The switch of the live traffic, mainly the hosts are used for the backend pools configuration at Application Gateway level, in this way the switch is transparent for the end user of the workloads.
+For the testing and validation purpose the hosts are also exposed at Application Gateway level with dedicated endpoints and also at AKS Ingress controller level to manage the TLS in the proper way.
+At this stage the validation is based on the infra and app monitoring metrics and official SLO and SLI, when avaialble. If the validation gate is satisfied than is possible to move in the [last state](#t4-blue-cluster-is-destroyed) of the pattern.
+
+
+#### T4: Blue cluster is destroyed
+The execution of the switch of the traffic bring the pattern in the final stage, in which there is still a validation and monitoring that the green cluster is working as expected also with live traffic; is always important to remember that the valadiation and monitoring cover both platform and application level.
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:119]: Consider removing "also" from the first sentence. Also add "It" in front of "is always important ..."

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+After this validation is completed, then the blue cluster can be destroyed.
+The destruction is a step that is strongly recommended in order to reduce costs and make the proper usage of the elastcity provided by Azure, in particular AKS.
+
+![Step4](../media/bg-step4.png)
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:123]: Aren't we destroying the blue app gateways as well?

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+
+
+### Components
+Below the main components and azure services that are part of the blue green deployment for AKS.
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:127]: add "are" after "Below"

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+- [Application Gateway](https://azure.microsoft.com/services/application-gateway/), the main responsability is to act as gateway and Load balancer for the AKS clusters.
+- [AKS](https://azure.microsoft.com/services/kubernetes-service/), is the subject of the patter.
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:129]: Change "patter" to "pattern"

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+- [Container Registry](https://azure.microsoft.com/services/container-registry/), has the main role to store and distribute the artifacts in the AKS clusters, example of artifacts are Container Images, HELM Charts, etc.. .
+- [Azure Monitor](https://azure.microsoft.com/services/monitor/), is the core monitoring platform for AKS, it is strongly recommended with his native integration with AKS, providing logs, metrics and alerting to manage the different stages of the pattern.
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:131]: Consider changing "is the core monitoring platform for AKS, it is strongly recommended with his native integration with AKS, providing logs, metrics and alerting to manage the different stages of the pattern" with "is the core observability platform for AKS and strongly recommended given its native integration with AKS and its ability to provide logging, monitoring and alerting capabilities used to manage the different stages of this pattern."

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+- [Azure Firewall](https://azure.microsoft.com/services/azure-firewall/), has the main responsability to manage the egress traffic into the clusters, in partcular in hub and spoke topology and connections with external datacenters or locations.
+- [KeyVault](https://azure.microsoft.com/services/key-vault/), it is required to manage in the proper way the secrets and certificates that need to be use at AKS level.
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:133]: Consider changing to: "is recommended for securely managing secrets and certificates used by the Azure resources and the applications that depend on them".

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+
+
+### Alternatives
+From a pattern perspective there altrenative scenarios to implement a more controlled switch between the cluster, that means that the core pattern remin the same, and the main change is on the traffic switch method, just as example is possible to have a *canary release* with traffic rules based on:
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:137]: There are spelling errors in this sentence and its needs to be revised for clarity sake.

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 22]
```diff
@@ -0,0 +1,222 @@
...
+- percentage of the incoming traffic
+- http headers
+- Cookies
+
+Another alternative that has more impact on the blast radius of the deployment is to have a ring based deployments. Instead of just blue and green clusters, it is possible to have more clusters called rings. Each ring is large enought for the number of users that have access to the new version/config of the AKS. As for the blue green pattern described, the rings can be removed to have the proper cost optimization and control.
+
+There are two Azure Servies listed in the [Componets Section](#components), for which is possible to use also alternative products and/or OSS solutions. The two services are:
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:144]: Consider changing to "There are two Azure Servies listed in the [Componets Section](#components), for which it is possible to use alternate 3rd party products or OSS solutions. These two services are:"

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 23]
```diff
@@ -0,0 +1,222 @@
...
+
+One important lesson learned is to have dedicated hostnames for the blue green clusters and also have a dedicated endpoints configuration on the Gateway/Load Balancer in front of the clusters. This is critical to improve the reliability and validity of the deployment of the new cluster. In this way the validation of the deployment happen with the same architecture and configurations of a standard production cluster.
+
+The blue green pattern is also an enabler to implement and test BC/DR solutions for AKS and related workloads.
+
+A key aspect is related to the **Cost Optimization**, in detail the blue green cluster pattern is widely adopted in the cloud for the native elasticiy provided by cloud providers, in particular Azure, this include the cost saving and control to manage and mantain the blue green cluster deploymentm; in paritcular as described the main action is to destroy the cluster that is no longer needed.
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:167]: Revise this sentence as it is not easy to understand

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 24]
```diff
@@ -0,0 +1,222 @@
...
+
+It is important to highlight that a succesfull implementation of the pattern is related to the fact that the all the aspects like automation, monitoring and validation need to be applied at AKS Platform levele but also the workloads/apps deployed on top of it; only with the E2E coverge is possible to really benefit of the value of the blue green pattern.
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:169]: Revise this sentence as it is not easy to understand

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 25]
```diff
@@ -0,0 +1,222 @@
...
+
+
+### Reliability
+The blue green pattern has a direct and positive impact on the availability of AKS platform and workloads. In particular the pattern improves the availability during the deployment of the AKS platform changes, in particular downtime is near to zero and it can be affected by how user sessions are managed. Moreover the blue green provide also covergae for reliability during the deployment because by default there is the option to fallback in the previous version of the AKS cluster.
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:173]: Change "covergae' to "coverage". Consider copying and pasting the entire page into Word to perform a spell check.
Also the last sentence needs to be revised as it is not clear what you mean.

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 26]
```diff
@@ -0,0 +1,222 @@
...
+Here more detail about the resiliency and avalaibility pillar defined in the [Well Architected Framework](/azure/architecture/framework/resiliency/overview).
+
+
+### Cost optimization
+
+As described before on of the main advantages and benefits of the blue green deployment is to mantain a control and optimization of the costs without impacting the resiliency, availability, continuous delivery of the workloads and apps. This is achieved with the automation of the destroyment of the old cluster after that the switch is completed and validated. Another important point to mention is that to cotinue to have the same cost baseline the two clusters are usually hosted in the same subnet, in this way all the network connections and access to the resources/services is the same, that means that all the azure services and resources remain the same during the blue green deployment.
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:179]: There are multiple spelling errors. Also consider rephrasing " If we should quantify the azure costs to implement this pattern we can say that you shoudl double the costs of the AKS services during the time of the blue green deployment, that usually happen in hours; this also another reason to explain the large adoption of this pattern" to  "The Azure costs of this pattern implementation can be estimated at double the costs of the AKS services during the time of the blue green deployment, which usually last only for a few hours"

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md [hunk 27]
```diff
@@ -0,0 +1,222 @@
...
+If we should quantify the azure costs to implement this pattern we can say that you shoudl double the costs of the AKS services during the time of the blue green deployment, that usually happen in hours; this also another reason to explain the large adoption of this pattern.
+If you are courios and want to have more insight about cost optmization you can [here](/azure/architecture/framework/cost/overview).
+
+
+### Operational excellence
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:184]: I don't think what you have in this paragraph addresses the Operational questions around the use of this pattern. From an operational perspective, what would the customer have to do extra or different when implementing this pattern? Do we need to implement and monitor additional alerts to trigger Blue/Green deployments?

File: Scenarios/BlueGreen-Deployment-for-AKS/README.md [hunk 28]
```diff
@@ -0,0 +1,11 @@
+# Blue Green Deployment for AKS
+This scenario is dedicated to the blue green deployment at AKS level, the main purpose is to provide a reliable and high availability solution when:
```
Comment (bahramr) [inline Scenarios/BlueGreen-Deployment-for-AKS/README.md:2]: Maybe revise to: This architectural pattern describes how to properly implement a Blue/Green deployment of an AKS cluster. For Blue GReen deployment at the application level, refer to [this article
](https://docs.microsoft.com/en-us/azure/architecture/example-scenario/blue-green-spring/blue-green-spring).

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 2 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:38]: rephrased in  "including the sequence of events to have the proper traffic switch, from networking and app perspcetve, between the blue and green clusters."

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 5 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:81]: What I want to say and I will rephrase, is that after the deployment of the new cluster is required to apply all the configurations, deployment of the application in order to have the parity with the existing cluster.
@bahramr

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 7 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:85]: added some examples and more xplanation. @bahramr

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 8 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:86]: added some examples and more xplanation. @bahramr

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 9 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:92]: fixed

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 10 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:98]: this is more related to the application itself and not the K8S state. I mean that if you have the concept of session or you have a stateful component like a database then is up to the application team define the proper strategy in order that blue green will not break from an end user perspective.
@bahramr

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 13 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:116]: fixed

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 15 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:119]: fixed

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 17 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:127]: fixed

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 18 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:129]: fixed

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 19 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:131]: done

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 20 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:133]: fixed

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 22 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:144]: fixed

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 25 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:173]: fixed

File: Scenarios/BlueGreen-Deployment-for-AKS/README.md (hunk 28 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/README.md:2]: fixed

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 1 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:13]: fixed

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 11 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:75]: restructured all images and diagrams.
@bahramr

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 26 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:179]: fixed all spelling errors, via word review
@bahramr

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 27 above)
Comment (bahramr) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md:185]: Consider reviewing the DevOps Checklist (https://docs.microsoft.com/en-us/azure/architecture/checklist/dev-ops) for considerations to incorporate here, around monitoring, testing, azure policies for example

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 23 above)
Comment (bahramr) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md]: Still need to be addressed

Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md]: This is also part of the Cost Optimization paragraph, here is just an high level introduction that is part of the  "Considerations" section.

Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md]: rephrased, which is the statement that is not clear?
@bahramr

File: Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md (hunk 24 above)
Comment (vimorra) [reply Scenarios/BlueGreen-Deployment-for-AKS/GAP/bluegreen-aks-solution-content.md]: Also the statement at line @160 is not clear?
@bahramr

**Pull request merged**